code/game/jobs/job_exp.dm

```diff
@@ -0,0 +1,211 @@
+		job_requirement = config.use_exp_restrictions_heads_hours * 60
+	if(my_exp >= job_requirement)
+		return 0
+	else
+		return (job_requirement - my_exp)
+
+
+
+/proc/job_is_xp_locked(jobtitle)
+	if(!config.use_exp_restrictions_heads && ((jobtitle in command_positions) || jobtitle == "AI"))
+		return 0
+	if(!config.use_exp_restrictions_other && !((jobtitle in command_positions) || jobtitle == "AI"))
+		return 0
+	return 1
+
+
+/mob/proc/get_exp_report()
+	if(client)
+		return client.get_exp_report()
+	else
+		return "[src] has no client."
+
+/client/proc/get_exp_report()
+	if(!config.use_exp_tracking)
+		return "Tracking is disabled in the server configuration file."
+
+	var/list/play_records = params2list(prefs.exp)
+	if(!play_records.len)
+		return "[key] has no records."
+
+	var/return_text = "<UL>"
+
+	var/list/exp_data = list()
+	for(var/cat in exp_jobsmap)
+		if(text2num(play_records[cat]))
+			exp_data[cat] = text2num(play_records[cat])
+		else
+			exp_data[cat] = 0
+
+	for(var/dep in exp_data)
+		if(exp_data[dep] > 0)
+			if(dep == EXP_TYPE_EXEMPT)
+				return_text += "<LI>Exempt (all jobs auto-unlocked)</LI>"
+			else if(exp_data[EXP_TYPE_CREW] > 0)
+				var/my_pc = num2text(round(exp_data[dep]/exp_data[EXP_TYPE_CREW]*100))
+				return_text += "<LI>[dep] [get_exp_format(exp_data[dep])] ([my_pc]%)</LI>"
+			else
+				return_text += "<LI>[dep] [get_exp_format(exp_data[dep])] </LI>"
+
+	return_text += "</UL>"
+	var/list/jobs_locked = list()
+	var/list/jobs_unlocked = list()
+	for(var/datum/job/job in job_master.occupations)
+		if(job.exp_requirements && job.exp_type)
+			if(!job_is_xp_locked(job.title))
+				continue
+			else if(!job.available_in_playtime(mob.client))
+				jobs_unlocked += job.title
+			else
+				var/xp_req = job.exp_requirements
+				if(config.use_exp_restrictions_heads_hours && ((job.title in command_positions) || job.title == "AI"))
+					xp_req = config.use_exp_restrictions_heads_hours * 60
+				jobs_locked += "[job.title] [get_exp_format(text2num(play_records[job.exp_type]))] / [get_exp_format(xp_req)] [job.exp_type] EXP)"
+	if(jobs_unlocked.len)
+		return_text += "<BR><BR>Jobs Unlocked:<UL> "
+		for(var/text in jobs_unlocked)
+			return_text += "<LI>[text]</LI>"
+		return_text += "</UL>"
+	if(jobs_locked.len)
+		return_text += "<BR><BR>Jobs Not Unlocked:<UL>"
+		for(var/text in jobs_locked)
+			return_text += "<LI>[text]</LI>"
+		return_text += "</UL>"
+	return return_text
+
+
+/client/proc/get_exp_general()
+	var/list/play_records = params2list(prefs.exp)
+	var/exp_gen = text2num(play_records[EXP_TYPE_CREW])
+	return get_exp_format(exp_gen)
+
+/proc/get_exp_format(var/expnum)
+	if(expnum > 60)
+		return num2text(round(expnum / 60)) + "h"
+	else if(expnum > 0)
+		return num2text(expnum) + "m"
+	else
+		return "0h"
+
+/proc/update_exp(var/minutes, var/announce_changes = 0)
+	if(!establish_db_connection())
+		return -1
+	spawn(0)
+		for(var/client/C in clients)
```

Comment:
Everything below here (except the sleep) should either be in a `try`/`catch` or another `spawn`, so if there's a runtime while trying to update exp for one client, later clients won't simply get skipped.

Instead of indenting all the code below again, you may want to simply move it to a new `/client/proc`, and have this proc call that one for each client.